src/update_handler.c

```diff
@@ -2527,6 +2532,102 @@ static gboolean check_if_area_is_clear(const gchar *device, guint64 start, gsize
 	return res;
 }
 
+static gboolean img_to_boot_nor_fallback_handler(RaucImage *image, RaucSlot *dest_slot, const gchar *hook_name, GError **error)
+{
+	GError *ierror = NULL;
+	gboolean res = FALSE;
+	gchar** devices = NULL;
+	gboolean primary_clear;
+	char *part_names[2] = {
+		"primary",
+		"fallback"
+	};
+	int first_part_desc_index = 1;
+	const gsize header_size = 512; /* TODO: this value should be settable through the system config file. */
+
+	g_return_val_if_fail(image, FALSE);
+	g_return_val_if_fail(dest_slot, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* run slot pre install hook if enabled */
+	if (hook_name && image->hooks.pre_install) {
+		res = run_slot_hook(hook_name, R_SLOT_HOOK_PRE_INSTALL, image, dest_slot, &ierror);
+		if (!res) {
+			g_propagate_error(error, ierror);
+			goto out;
+		}
+	}
+
+	devices = g_strsplit(dest_slot->device, ":", -1);
+	if (g_strv_length(devices) != 2) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Slot device string does not contain two comma separated devices. Aborting!");
+		goto out;
+	}
+
+	g_message("Executing nor fallback handler on slot %s with primary device %s and fallback device %s",
+			dest_slot->name, devices[0], devices[1]);
+
+	/* If the primary partition is not fully programmed, it most likely means that the fallback
+	 * partition was used to boot and is therefore valid. To avoid ending up with two broken partitions,
+	 * upgrade the primary partition first.
+	 *
+	 * We are only checking the header area, as it is the last thing written.
+	 */
+	res = check_if_area_is_clear(devices[0], 0, header_size, &primary_clear, &ierror);
```

Comment:
Compared to block devices, NOR flash doesn't have atomic writes for 512 byte blocks. That results in some additional cases that need to be handled.

- Power loss while writing the primary header (after successfully writing the secondary image)
- Primary header is incomplete, but not clear
- Secondary image is used for booting
- Second update attempt erases secondary image, as primary header is not clear ⇒ temporarily no valid image

This means we'd need a better check to detect if the header is incomplete/corrupt. For simple cases, it might be enough if the last byte of the header region is != 0.

- Power loss while writing the primary header (after successfully writing the secondary image)
- Primary header is incomplete, but seems valid to the ROM code
- Boot fails

This depends on the SoC: For headers with a proper checksum, this shouldn't be a problem. If not it probably needs SoC specific research on how it actually behaves for partially written headers.


An alternative solution would be to keep some state in the RAUC data partition, to be able to detect the different cases.

If these corner-cases are acceptable in your use case, I'd also be OK with only the current approach of checking if the header is clear. In that case, there needs to be a big warning in the documentation, though.